### PR TITLE
[FIXED JENKINS-37223]

### PR DESCRIPTION
**Pull request by kohsuke**
Added a simple diagnostic HTTP response to the TcpSlaveAgentListener.

### CODE REVIEW

File: core/src/main/java/hudson/TcpSlaveAgentListener.java [hunk 1]
```diff
@@ -235,6 +252,48 @@ public void run() {
...
+                    boolean first = true;
+                    for (AgentProtocol p : AgentProtocol.all()) {
+                        if (first)  first = false;
+                        else        o.write(",");
+                        o.write(p.getName());
+                    }
+                    o.write("\n");
+                    o.write("Jenkins-Version: " + Jenkins.VERSION + "\n");
+                    o.write("Jenkins-Session: " + Jenkins.SESSION_HASH + "\n");
+                    o.write("Client: " + s.getInetAddress().getHostAddress() + "\n");
```
Comment (jtnord) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: You are printing "Client" and giving the servers address.

perhaps you want to print "Server"?

File: core/src/main/java/hudson/TcpSlaveAgentListener.java [hunk 2]
```diff
@@ -235,6 +252,48 @@ public void run() {
...
+        /**
+         * Respond to HTTP request with simple diagnostics.
+         * Primarily used to test the low-level connectivity.
+         */
+        private void respondHello(String header, Socket s) throws IOException {
+            try {
+                Writer o = new OutputStreamWriter(s.getOutputStream(), "UTF-8");
+
+                if (header.startsWith("GET / ")) {
+                    o.write("HTTP/1.0 200 OK\n");
```
Comment (jtnord) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: :bug: HTTP mandates the use of '\r\n'

**Comment by jtnord** [comment]
looks good - but the response violates the HTTP spec, and a confusing address will be printed for the "Client"

File: core/src/main/java/hudson/TcpSlaveAgentListener.java (hunk 1 above)
Comment (kohsuke) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: This is correctly giving out the IP address of the connecting client that the server is seeing.

See [Socket#getInetAddress() javadoc](https://docs.oracle.com/javase/7/docs/api/java/net/Socket.html#getInetAddress()), and the following experiments:

```
% groovysh
Groovy Shell (2.0.2, JVM: 1.7.0_07)
Type 'help' or '\h' for help.
--------------------------------------------------------------------------------------------------------------------------------------------
groovy:000> s=new Socket("kohsuke.org",80)
===> Socket[addr=kohsuke.org/208.113.153.110,port=80,localport=50380]
groovy:000> s.inetAddress
===> kohsuke.org/208.113.153.110
groovy:000> 
```

Comment (jtnord) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: OK my mistake - I was confusing this with `ServerSocket`
I would recommend that you should print the local address as well so you can diagnose on the client where it actually connected to (when it goes via a load balancer)

**Comment by jtnord** [comment]
:bee: :+1:

File: core/src/main/java/hudson/TcpSlaveAgentListener.java [hunk 3]
```diff
@@ -235,6 +252,49 @@ public void run() {
...
+                    o.write("\r\n");
+                    o.write("Jenkins-Agent-Protocols: ");
+                    boolean first = true;
+                    for (AgentProtocol p : AgentProtocol.all()) {
+                        if (first)  first = false;
+                        else        o.write(",");
+                        o.write(p.getName());
+                    }
+                    o.write("\r\n");
+                    o.write("Jenkins-Version: " + Jenkins.VERSION + "\r\n");
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: Smells like a code duplication from somewhere

**Pull request merged**